### PR TITLE
[workflow] Fix the object loss due to driver exit issues.

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?

When the workflow runs in driver mode, the owner of the object ref is the driver. So when the driver exits, the objects are no longer available. This happens when we run with `run_async`.

This PR fixed this by passing the manager actor as the owner of the objects.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by suquark — COMMENTED** [review]
~~we should get the workflow actor manager at the beginning of the function instead, so we won't get the actors repeatedly for every object~~

**Comment by suquark** [comment]
Also could you verify if this works under client mode?

File: python/ray/workflow/workflow_state_from_dag.py [hunk 1]
```diff
@@ -168,7 +168,10 @@ def _node_visitor(node: Any) -> Any:
                     flattened_args = _SerializationContextPreservingWrapper(
                         flattened_args
                     )
-                input_placeholder: ray.ObjectRef = ray.put(flattened_args)
+                workflow_manager = workflow_access.get_management_actor()
```
Comment (suquark) [inline python/ray/workflow/workflow_state_from_dag.py:171]: we should get the workflow actor manager at the beginning of the function instead, so we won't get the actors repeatedly for every object

**Comment by iycheng** [comment]
@suquark the ray client is supported.

@ckw017 could you take a look at the ray client updates?

**Review by ckw017 — COMMENTED** [review]
Client part looks good

File: src/ray/protobuf/ray_client.proto [hunk 2]
```diff
@@ -114,6 +114,8 @@ message PutRequest {
   int32 total_chunks = 4;
   // Total size in bytes of the data being put
   int64 total_size = 5;
+  // The owner of the put
+  bytes _owner_id = 6;
```
Comment (ckw017) [inline src/ray/protobuf/ray_client.proto:118]: nit: any reason why this needs to be prefixed with underscore?

Comment (iycheng) [reply src/ray/protobuf/ray_client.proto:118]: I'm just following the non-client mode API:

https://github.com/ray-project/ray/blob/master/python/ray/_private/worker.py#L2314

and it's ray.put(_owner=xxx). But this is protobuf, I think I can change it to the one without `_`. Up to you.

Comment (ckw017) [reply src/ray/protobuf/ray_client.proto:118]: If you can drop from the protobuf that would be nice, I think there's some [wonky behavior](https://github.com/golang/protobuf/issues/201) sometimes when you prefix with non-alphabetical characters.

**Review by suquark — APPROVED** [review]
LGTM.

**Review by ckw017 — APPROVED** [review]

**Comment by SebastianMorawiec** [comment]
Hey @iycheng - sorry for jumping in, but I'd understand that this change would fix this issue Ive reported: https://github.com/ray-project/ray/issues/29253, am I right? Is there a plan to merge this PR?

**Comment by iycheng** [comment]
@SebastianMorawiec sorry I missed your message. The root cause is because of ownership. In ray the one who create the object own the object. Here it's the driver. And when the owner died (driver exits), the object will be lost.

The fix is to make the manager as the owner and thus it'll always be there until no one is using that.

**Pull request merged**